substrate/client/network/types/src/kad.rs

```diff
@@ -0,0 +1,171 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{multihash::Multihash, PeerId};
+use bytes::Bytes;
+use litep2p::protocol::libp2p::kademlia::{Record as LiteRecord, RecordKey};
+use std::{borrow::Borrow, error::Error, fmt, time::Instant};
+
+/// The (opaque) key of a record.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Key(Bytes);
+
+impl Key {
+	/// Creates a new key from the bytes of the input.
+	pub fn new<K: AsRef<[u8]>>(key: &K) -> Self {
+		Key(Bytes::copy_from_slice(key.as_ref()))
+	}
+
+	/// Copies the bytes of the key into a new vector.
+	pub fn to_vec(&self) -> Vec<u8> {
+		Vec::from(&self.0[..])
+	}
+}
+
+impl Borrow<[u8]> for Key {
+	fn borrow(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl AsRef<[u8]> for Key {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl From<Vec<u8>> for Key {
+	fn from(v: Vec<u8>) -> Key {
+		Key(Bytes::from(v))
+	}
+}
+
+impl From<Multihash> for Key {
+	fn from(m: Multihash) -> Key {
+		Key::from(m.to_bytes())
+	}
+}
+
+/// A record stored in the DHT.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Record {
+	/// Key of the record.
+	pub key: Key,
+	/// Value of the record.
+	pub value: Vec<u8>,
+	/// The (original) publisher of the record.
+	pub publisher: Option<PeerId>,
+	/// The expiration time as measured by a local, monotonic clock.
+	pub expires: Option<Instant>,
+}
+
+impl Record {
+	/// Creates a new record for insertion into the DHT.
+	pub fn new<K>(key: K, value: Vec<u8>) -> Self
+	where
+		K: Into<Key>,
```

Comment:
This is likely why extra template arguments are needed. I would just make it:
```
pub fn new(key: Key, value: Vec<u8>) -> Self
```
And called `.into()` explicitly.